File: go/varint.go
package veriform

import (
	"bytes"
	"encoding/binary"
	"errors"
	"io"
)

// EncodeVarint encodes a uint64 into buf and returns the number of bytes written.
// If the buffer is too small, EncodeVarint will panic.
func EncodeVarint(buf []byte, input uint64) int {
	length := 1
	value := (input << 1) | 1
	max := uint64(1 << 7)

	for input >= max {
		// 9-byte special case
		if length == 8 {
			buf[0] = 0
			binary.LittleEndian.PutUint64(buf[1:9], input)
			return 9
		}

		value <<= 1
		max <<= 7
		length++
	}

	var output [8]byte
	binary.LittleEndian.PutUint64(output[:], value)
	copy(buf, output[:length])
	return length
}

func DecodeVarint(r io.Reader) (uint64, error) {
	var result uint64
	var buf [8]byte

	_, err := r.Read(buf[:1])
	if err != nil {
		return 0, err
	}

	prefix := buf[0]
	if prefix == 0 {
		err := binary.Read(r, binary.LittleEndian, &result)

		if err == nil && result < (1<<56) {
			return 0, errors.New("malformed varint")
		}

		return result, err
	}

	length := uint(1)

	// TODO: use math/bits TrailingZeros() if/when it becomes available
	// See: https://github.com/golang/go/issues/18616
	for prefix&1 == 0 {
		length += 1
		prefix >>= 1
	}

	_, err = io.ReadFull(r, buf[1:length])
	if err != nil {
		return 0, err
	}

	err = binary.Read(bytes.NewReader(buf[:]), binary.LittleEndian, &result)
	if err != nil {
		return 0, err
	}

	result >>= length
	if length > 1 && result < (1<<(7*(length-1))) {
		return 0, errors.New("malformed varint")
	}

	return result, nil
}

File: rust/src/varint.rs
//! Variable-width 64-bit little endian integers


#[cfg(not(feature = "std"))]
use alloc::string::ToString;
use byteorder::{ByteOrder, LittleEndian};
use errors::*;

/// Encode a 64-bit unsigned integer in zsuint64 form
///
/// Panics if the `out` slice is not large enough to store the result.
///
/// Recommended size of `out`: minimum of 9 bytes to ensure panic-free
/// operation for all inputs.
pub fn encode(value: u64, out: &mut [u8]) -> usize {
    let mut length = 1;
    let mut result = (value << 1) | 1;
    let mut max = 1 << 7;

    while value >= max {
        // 9-byte special case
        if length == 8 {
            out[0] = 0;
            LittleEndian::write_u64(&mut out[1..9], value);
            return 9;
        }

        result <<= 1;
        max <<= 7;
        length += 1;
    }

    LittleEndian::write_uint(out, result, length);
    length
}

/// Decode a zsuint64-encoded unsigned 64-bit integer
pub fn decode(input: &mut &[u8]) -> Result<u64> {
    let bytes = *input;

    let prefix = *bytes.first().ok_or_else(|| {
        ErrorKind::TruncatedMessage("missing varint prefix".to_string())
    })?;

    if prefix == 0 {
        if bytes.len() >= 9 {
            let result = LittleEndian::read_u64(&bytes[1..9]);

            if result < (1 << 56) {
                return Err(
                    ErrorKind::CorruptedMessage("malformed varint".to_string()).into(),
                );
            }

            *input = &bytes[9..];
            return Ok(result);
        } else {
            return Err(
                ErrorKind::TruncatedMessage("truncated varint".to_string()).into(),
            );
        }
    }

    let length = prefix.trailing_zeros() as usize + 1;

    if bytes.len() < length {
        return Err(
            ErrorKind::TruncatedMessage("truncated varint".to_string()).into(),
        );
    }

    let result = LittleEndian::read_uint(bytes, length) >> length;

    if length > 1 && result < (1 << (7 * (length - 1))) {
        return Err(
            ErrorKind::CorruptedMessage("malformed varint".to_string()).into(),
        );
    }

    *input = &bytes[length..];
    Ok(result)
}

#[cfg(feature = "bench")]
mod bench {
    #[cfg(feature = "bench")]
    use leb128;
    #[cfg(feature = "bench")]
    use test::Bencher;
    use varint;

    #[cfg(feature = "bench")]
    #[bench]
    fn bench_encode(b: &mut Bencher) {
        let mut output = [0u8; 9];

        // 2**48 + 31337
        b.iter(|| varint::encode(281474976741993, &mut output));
    }

    #[cfg(feature = "bench")]
    #[bench]
    fn bench_decode(b: &mut Bencher) {
        let input = b"\xC04=\x00\x00\x00\x80";

        // 2**48 + 31337
        b.iter(|| {
            let mut readable = &input[..];
            varint::decode(&mut readable).unwrap()
        });
    }

    #[cfg(feature = "bench")]
    #[bench]
    fn bench_leb128_encode(b: &mut Bencher) {
        let mut output = [0u8; 9];

        // 2**48 + 31337
        b.iter(|| {
            let mut writable = &mut output[..];
            leb128::write::unsigned(&mut writable, 281474976741993).unwrap()
        });
    }

    #[cfg(feature = "bench")]
    #[bench]
    fn bench_leb128_decode(b: &mut Bencher) {
        let input = b"\xE9\xF4\x81\x80\x80\x80@";

        // 2**48 + 31337
        b.iter(|| {
            let mut readable = &input[..];
            leb128::read::unsigned(&mut readable).unwrap()
        });
    }
}
